File: client/client/rpc/protocol.py
import struct
import socket

from client.rpc.types import Type


class ProtocolBase(object):
    def __init__(self, trans):
        self.trans = trans

    def write_message_begin(self, name, typeid, seqid):
        raise NotImplementedError

    def write_message_end(self):
        raise NotImplementedError

    def write_field_begin(self, name, typeid, fid):
        raise NotImplementedError

    def write_field_end(self):
        raise NotImplementedError

    def write_field_stop(self):
        raise NotImplementedError

    def write_list_begin(self, elemtype, size):
        raise NotImplementedError

    def write_list_end(self):
        raise NotImplementedError

    def write_bool(self, value):
        raise NotImplementedError

    def write_byte(self, value):
        raise NotImplementedError

    def write_i16(self, value):
        raise NotImplementedError

    def write_i32(self, value):
        raise NotImplementedError

    def write_float(self, value):
        raise NotImplementedError

    def write_string(self, value):
        raise NotImplementedError

    def write_binary(self, value):
        raise NotImplementedError

    def read_message_begin(self):
        raise NotImplementedError

    def read_message_end(self):
        raise NotImplementedError

    def read_field_begin(self):
        raise NotImplementedError

    def read_field_end(self):
        raise NotImplementedError

    def read_bool(self):
        raise NotImplementedError

    def read_byte(self):
        raise NotImplementedError

    def read_i16(self):
        raise NotImplementedError

    def raise_i32(self):
        raise NotImplementedError

    def read_float(self):
        raise NotImplementedError

    def read_string(self):
        raise NotImplementedError

    def read_binary(self):
        raise NotImplementedError
    
    def read_list_begin(self):
        raise NotImplementedError

    def read_list_end(self):
        raise NotImplementedError


class BinaryProtocol(ProtocolBase):
    def __init__(self, trans):
        super().__init__(trans)

    def write_message_begin(self, name, typeid, seqid):
        self.write_string(name)
        self.write_byte(typeid)
        self.write_i32(seqid)

    def write_message_end(self):
        pass

    def write_field_begin(self, name, typeid, fid):
        self.write_byte(typeid)
        self.write_i16(fid)

    def write_field_end(self):
        pass

    def write_field_stop(self):
        self.write_byte(Type.STOP)

    def write_list_begin(self, elemtype, size):
        self.write_byte(elemtype)
        self.write_i32(size)

    def write_list_end(self):
        pass

    def write_bool(self, value):
        self.write_byte(1 if value else 0)

    def write_byte(self, value):
        buf = struct.pack("!b", value)
        self.trans.write(buf)

    def write_i16(self, value):
        buf = struct.pack("!h", value)
        self.trans.write(buf)

    def write_i32(self, value):
        buf = struct.pack("!i", value)
        self.trans.write(buf)

    def write_float(self, value):
        buf = struct.pack("!f", value)
        self.trans.write(buf)

    def write_string(self, value):
        self.write_binary(bytes(value, "utf8"))

    def write_binary(self, value):
        self.write_i32(len(value))
        self.trans.write(value)

    def read_message_begin(self):
        size = self.read_i32()
        name = self.trans.read_all(size)
        typeid = self.read_byte()
        seqid = self.read_i32()

        return (name, typeid, seqid)

    def read_message_end(self):
        pass

    def read_field_begin(self):
        typeid = self.read_byte()
        if typeid == Type.STOP:
            return (None, typeid, 0)
        fid = self.read_i16()
        return (None, typeid, fid)

    def read_field_end(self):
        pass
    
    def read_list_begin(self):
        etype = self.read_byte()
        size = self.read_i32()
        return (etype, size)

    def read_list_end(self):
        pass

    def read_bool(self):
        byte = self.read_byte()
        return not not byte

    def read_byte(self):
        buf = self.trans.read_all(1)
        val, = struct.unpack("!b", buf)
        return val

    def read_i16(self):
        buf = self.trans.read_all(2)
        val, = struct.unpack("!h", buf)
        return val

    def read_i32(self):
        buf = self.trans.read_all(4)
        val, = struct.unpack("!i", buf)
        return val

    def read_float(self):
        buf = self.trans.read_all(4)
        val, = struct.unpack("!f", buf)
        return val

    def read_string(self):
        return self.read_binary().decode("utf8")

    def read_binary(self):
        size = self.read_i32()
        s = self.trans.read_all(size)
        return s

File: server/rpc/protocol.go
package rpc

import (
	"bytes"
	"encoding/binary"
	"errors"
	"io"
	"math"
)

// Interfaces

type ProtocolFactory interface {
	GetProtocol(trans Transport) Protocol
}

type Protocol interface {
	WriteMessageBegin(name string, typeID MessageType, seqID int32) error
	WriteMessageEnd() error
	WriteFieldBegin(name string, typeID Type, id int16) error
	WriteFieldEnd() error
	WriteFieldStop() error
	WriteListBegin(elemType Type, size int) error
	WriteListEnd() error
	WriteBool(value bool) error
	WriteByte(value byte) error
	WriteI16(value int16) error
	WriteI32(value int32) error
	WriteFloat(value float32) error
	WriteString(value string) error
	WriteBinary(value []byte) error

	ReadMessageBegin() (name string, typeID MessageType, seqID int32, err error)
	ReadMessageEnd() error
	ReadFieldBegin() (name string, typeID Type, id int16, err error)
	ReadFieldEnd() error
	ReadListBegin() (elemType Type, size int, err error)
	ReadListEnd() error
	ReadBool() (value bool, err error)
	ReadByte() (value byte, err error)
	ReadI16() (value int16, err error)
	ReadI32() (value int32, err error)
	ReadFloat() (value float32, err error)
	ReadString() (value string, err error)
	ReadBinary() (value []byte, err error)

	Flush() (err error)
	Transport() Transport
}

// Concrete binary protocol

type BinaryProtocol struct {
	// trans         TRichTransport
	trans *UDPSocket
	// reader        io.Reader
	// writer        io.Writer
	// strictRead  bool
	// strictWrite bool
	buffer [64]byte
}

type BinaryProtocolFactory struct{}

func NewBinaryProtocol(trans Transport) *BinaryProtocol {
	p := &BinaryProtocol{trans: trans.(*UDPSocket)}
	return p
}

func NewBinaryProtocolFactory() *BinaryProtocolFactory {
	return &BinaryProtocolFactory{}
}

func (p *BinaryProtocolFactory) GetProtocol(trans Transport) Protocol {
	return NewBinaryProtocol(trans)
}

// Write methods

func (p *BinaryProtocol) WriteBool(value bool) error {
	if value {
		return p.WriteByte(1)
	}
	return p.WriteByte(0)
}

func (p *BinaryProtocol) WriteByte(value byte) error {
	e := p.trans.WriteByte(value)
	return NewProtocolException(e)
}

func (p *BinaryProtocol) WriteI16(value int16) error {
	v := p.buffer[0:2]
	binary.BigEndian.PutUint16(v, uint16(value))
	_, e := p.trans.Write(v)
	return NewProtocolException(e)
}

func (p *BinaryProtocol) WriteI32(value int32) error {
	v := p.buffer[0:4]
	binary.BigEndian.PutUint32(v, uint32(value))
	_, e := p.trans.Write(v)
	return NewProtocolException(e)
}

func (p *BinaryProtocol) WriteFloat(value float32) error {
	return p.WriteI32(int32(math.Float32bits(value)))
}

func (p *BinaryProtocol) WriteString(value string) error {
	e := p.WriteI32(int32(len(value)))
	if e != nil {
		return NewProtocolException(e)
	}
	_, e = p.trans.WriteString(value)
	return NewProtocolException(e)
}

func (p *BinaryProtocol) WriteBinary(value []byte) error {
	e := p.WriteI32(int32(len(value)))
	if e != nil {
		return e
	}
	_, e = p.trans.Write(value)
	return NewProtocolException(e)
}

func (p *BinaryProtocol) WriteListBegin(elemType Type, size int) error {
	e := p.WriteByte(byte(elemType))
	if e != nil {
		return e
	}
	e = p.WriteI32(int32(size))
	return e
}

func (p *BinaryProtocol) WriteListEnd() error {
	return nil
}

func (p *BinaryProtocol) WriteFieldBegin(name string, typeID Type, id int16) error {
	e := p.WriteByte(byte(typeID))
	if e != nil {
		return e
	}
	e = p.WriteI16(id)
	return e
}

func (p *BinaryProtocol) WriteFieldEnd() error {
	return nil
}

func (p *BinaryProtocol) WriteFieldStop() error {
	e := p.trans.WriteByte(byte(Stop))
	return e
}

func (p *BinaryProtocol) WriteMessageBegin(name string, typeID MessageType, seqID int32) error {
	e := p.WriteString(name)
	if e != nil {
		return e
	}
	e = p.WriteByte(byte(typeID))
	if e != nil {
		return e
	}
	e = p.WriteI32(seqID)
	return e
}

func (p *BinaryProtocol) WriteMessageEnd() error {
	return nil
}

// Read methods

func (p *BinaryProtocol) ReadMessageBegin() (name string, typeID MessageType, seqID int32, err error) {
	size, e := p.ReadI32()
	if e != nil {
		return "", typeID, 0, NewProtocolException(e)
	}
	if size < 0 {
		name, e = p.ReadString()
		if e != nil {
			return name, typeID, seqID, NewProtocolException(e)
		}
		seqID, e = p.ReadI32()
		if e != nil {
			return name, typeID, seqID, NewProtocolException(e)
		}
		return name, typeID, seqID, nil
	}
	name, e2 := p.readStringBody(size)
	if e2 != nil {
		return name, typeID, seqID, e2
	}
	b, e3 := p.ReadByte()
	if e3 != nil {
		return name, typeID, seqID, e3
	}
	typeID = MessageType(b)
	seqID, e4 := p.ReadI32()
	if e4 != nil {
		return name, typeID, seqID, e4
	}
	return name, typeID, seqID, nil
}

func (p *BinaryProtocol) ReadMessageEnd() error {
	return nil
}

func (p *BinaryProtocol) ReadFieldBegin() (name string, typeID Type, seqID int16, err error) {
	t, err := p.ReadByte()
	typeID = Type(t)
	if err != nil {
		return name, typeID, seqID, err
	}
	if Type(t) != Stop {
		seqID, err = p.ReadI16()
	}
	return name, typeID, seqID, err
}

func (p *BinaryProtocol) ReadFieldEnd() error {
	return nil
}

var invalidDataLength = NewProtocolExceptionWithType(InvalidDataID, errors.New("Invalid data length"))

func (p *BinaryProtocol) ReadListBegin() (elemType Type, size int, err error) {
	b, e := p.ReadByte()
	if e != nil {
		err = NewProtocolException(e)
		return
	}
	elemType = Type(b)
	size32, e := p.ReadI32()
	if e != nil {
		err = NewProtocolException(e)
		return
	}
	if size32 < 0 {
		err = invalidDataLength
		return
	}
	size = int(size32)

	return
}

func (p *BinaryProtocol) ReadListEnd() error {
	return nil
}

func (p *BinaryProtocol) ReadBool() (bool, error) {
	b, e := p.ReadByte()
	v := true
	if b != 1 {
		v = false
	}
	return v, e
}

func (p *BinaryProtocol) ReadByte() (byte, error) {
	v, err := p.trans.ReadByte()
	return v, err
}

func (p *BinaryProtocol) ReadI16() (value int16, err error) {
	buf := p.buffer[0:2]
	err = p.readAll(buf)
	value = int16(binary.BigEndian.Uint16(buf))
	return value, err
}

func (p *BinaryProtocol) ReadI32() (value int32, err error) {
	buf := p.buffer[0:4]
	err = p.readAll(buf)
	value = int32(binary.BigEndian.Uint32(buf))
	return value, err
}

func (p *BinaryProtocol) ReadFloat() (value float32, err error) {
	buf := p.buffer[0:4]
	err = p.readAll(buf)
	value = math.Float32frombits(binary.BigEndian.Uint32(buf))
	return value, err
}

func (p *BinaryProtocol) ReadString() (value string, err error) {
	size, e := p.ReadI32()
	if e != nil {
		return "", e
	}
	if size < 0 {
		err = invalidDataLength
		return
	}

	return p.readStringBody(size)
}

func (p *BinaryProtocol) ReadBinary() ([]byte, error) {
	size, e := p.ReadI32()
	if e != nil {
		return nil, e
	}
	if size < 0 {
		return nil, invalidDataLength
	}

	isize := int(size)
	buf := make([]byte, isize)
	_, err := io.ReadFull(p.trans, buf)
	return buf, NewProtocolException(err)
}

func (p *BinaryProtocol) readAll(buf []byte) error {
	_, err := io.ReadFull(p.trans, buf)
	return NewProtocolException(err)
}

const readLimit = 32768

func (p *BinaryProtocol) readStringBody(size int32) (value string, err error) {
	if size < 0 {
		return "", nil
	}

	var (
		buf bytes.Buffer
		e   error
		b   []byte
	)

	switch {
	case int(size) <= len(p.buffer):
		b = p.buffer[:size] // avoids allocation for small reads
	case int(size) < readLimit:
		b = make([]byte, size)
	default:
		b = make([]byte, readLimit)
	}

	for size > 0 {
		_, e = io.ReadFull(p.trans, b)
		buf.Write(b)
		if e != nil {
			break
		}
		size -= readLimit
		if size < readLimit && size > 0 {
			b = b[:size]
		}
	}
	return buf.String(), NewProtocolException(e)
}

func (p *BinaryProtocol) Flush() (err error) {
	return NewProtocolException(p.trans.Flush())
}

func (p *BinaryProtocol) Transport() Transport {
	return p.trans
}
